src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -1490,17 +1490,19 @@ static QVariant fcnWordwrap( const QVariantList &values, const QgsExpressionCont
 static QVariant fcnLength( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
 {
   // two variants, one for geometry, one for string
-  if ( values.at( 0 ).userType() == qMetaTypeId< QgsGeometry>() )
+
+  //geometry variant
+  QgsGeometry geom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
+  if ( !geom.isNull() )
   {
-    //geometry variant
-    QgsGeometry geom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
-    if ( geom.type() != Qgis::GeometryType::Line )
+    if ( geom.type() == Qgis::GeometryType::Line )
+      return QVariant( geom.length() );
+    else
       return QVariant();
-
-    return QVariant( geom.length() );
   }
 
   //otherwise fall back to string variant
+  parent->setEvalErrorString( QString() );  // clear error string saying geometry not found
```

Comment:
Instead of this, I'd add a `bool tolerant = false` argument to QgsExpressionUtils::getGeometry, and if `tolerant` is true then don't set the eval error string in getGeometry